docs/concepts/native-interop.rst

```diff
@@ -0,0 +1,254 @@
+    namespace ConsoleApplication1 {
+
+        class Program {
+
+            delegate bool EnumWC(IntPtr hwnd, IntPtr lParam);
+
+            [DllImport("user32.dll")]
+            static extern int EnumWindows(EnumWC hWnd, IntPtr lParam);
+
+            static bool OutputWindow(IntPtr hwnd, IntPtr lParam) {
+                Console.WriteLine(hwnd.ToInt64());
+                return true;
+            }
+
+            static void Main(string[] args) {
+                EnumWindows(OutputWindow, IntPtr.Zero);
+            }
+        }
+    }
+    
+
+Before we walk through our example, it is good to go over the signatures of the 
+unmanaged functions we need to work with. The function we want to call to 
+enumerate all of the windows has the following signature:
+``BOOL EnumWindows (WNDENUMPROC lpEnumFunc, LPARAM lParam);``
+
+The first parameter is a callback. The said callback has the following signature: 
+``BOOL CALLBACK EnumWindowsProc (HWND hwnd, LPARAM lParam);``
+
+With this in mind, let's walk through the example:
+
+* Line #8 in the example defines a delegate that matches the signature of the 
+  callback from unmanaged code. Notice how the LPARAM and HWND types are 
+  represented using ``IntPtr`` in the managed code. 
+* Lines #10 and #11 introduce the ``EnumWindows`` function from the user32.dll 
+  library. 
+* Lines #13 - 16 implement the delegate. For this simple example, we just want 
+  to output the handle to the console.
+* Finally, in line #19 we invoke the external method and pass in the delegate.
+
+Both of the above examples depend on parameters, and in both cases, the parameters 
+are given as managed types. Runtime does the "right thing" and processes these 
+into its equivalents on the other side. Since this process is really important 
+to writing quality native interop code, let's take a look at what happens when 
+the runtime *marshals* the types. 
+
+Type marshalling
+^^^^^^^^^^^^^^^^
+**Marshalling** is the process of packing up types when they need to cross the 
+managed boundary into native and vice versa. 
+
+Why is this needed? Simply put, the types in the managed and unmanaged world 
+are different. In managed world, for instance, you have a ``String``, while in 
+the unmanaged world strings can be Unicode ("wide"), non-Unicode, null-terminated, 
+ASCII, etc. By default, the .NET runtime will try to do the Right Thing and for
+many applications leaving it to its own devices is usually fine. 
+
+However, for those situations where you need extra control, you can employ the 
+``MarshalAs`` attribute to tell the runtime what is the expected type in the 
+unmanaged world. For instance, if we want the string to be sent as a null-terminated 
+ANSI string, we could do it like this:
+
+.. code-block:: c#
+    
+    [DllImport("somenativelibrary.dll"]
+    static extern int MethodA([MarshalAs(UnmanagedType.LPStr) string parameter);
+
+Another aspect of type marshalling is how to pass in a struct to an unmanaged method.
```

Comment:
It seems like you are moving to a logically different topic, or sub-topic. It makes the text easier to read if you create a heading. Otherwise, it seems like a long block of undifferentiated text. It makes it much easier for the reader to see what they are looking for (or not, as the case may be).